File: src/main/python/Day16.py
import Utils as Utils

def sum_packet_versions(packet):
    return packet['version'] + sum(sum_packet_versions(p) for p in packet['packets'])

def calculate_expression(packet):
    result = 0
    type_id = packet['type_id']
    result = packet['value'] if type_id == 4 else calculate_expression(packet['packets'][0])
    for x in range(1, len(packet['packets'])):
        value = calculate_expression(packet['packets'][x])
        match type_id:
            case 0: result += value
            case 1: result *= value
            case 2: result = min(result, value)
            case 3: result = max(result, value)
            case 5: result = 1 if result > value else 0
            case 6: result = 1 if result < value else 0
            case 7: result = 1 if result == value else 0
    return result

def read_n(packetIter, count):
    return ''.join(next(packetIter) for x in range(count))

def read_packet(packetIter):
    packet = dict()
    packet['version'] = int(read_n(packetIter, 3), 2)
    packet['type_id'] = int(read_n(packetIter, 3), 2)
    packets = []
    packet['packets'] = packets
    
    if packet['type_id'] == 4:
        sb = ''
        while True:
            pack = read_n(packetIter, 5)
            sb += pack[1:]
            if pack[0] == '0':
                packet['value'] = int(sb, 2)
                break
    else:
        if next(packetIter) == '0':
            length = int(read_n(packetIter, 15), 2)
            packetSubIter = iter(read_n(packetIter, length))
            while True:
                try:
                    packets.append(read_packet(packetSubIter))
                except:
                    break
        else:
            length = int(read_n(packetIter, 11), 2)
            for x in range(length):
                packets.append(read_packet(packetIter))
    return packet

def hex_to_bin(line):
    return ''.join(('%6s' % bin(int(x, 16))).replace('0b', '').replace(' ', '0') for x in line)

@Utils.time_it_and_evaluate
def partOne(packet):
    return sum_packet_versions(packet)

@Utils.time_it_and_evaluate
def partTwo(packet):
    return calculate_expression(packet)

if __name__ == '__main__':
    line = Utils.readAll(2021, 'day16').strip()
    packet = read_packet(iter(hex_to_bin(line)))
    
    print('Day 16')
    partOne(packet)
    partTwo(packet)

File: src/main/java/hardcoded/aoc2021/Day16.java
package hardcoded.aoc2021;

import java.io.ByteArrayInputStream;
import java.io.DataInputStream;
import java.util.ArrayList;
import java.util.List;

import hardcoded.util.Utils;

public class Day16 {
	private static class Packet {
		List<Packet> packets = new ArrayList<>();
		int version;
		int typeId;
		long value;
	}
	
	public static void main(String[] args) throws Exception {
		String line = Utils.readAll(2021, "day16").strip();
		
		String binary = String.join("", line.chars()
			.mapToObj(c -> "%4s".formatted(Integer.toBinaryString(Character.digit(c, 16))).replace(' ', '0')).toList());
		Packet packet = readPacket(new DataInputStream(new ByteArrayInputStream(binary.getBytes())));
		
		Utils.printf("Day 16\n");
		Utils.printf("PartOne: %d\n", partOne(packet));
		Utils.printf("PartTwo: %d\n", partTwo(packet));
	}
	
	public static Packet readPacket(DataInputStream stream) throws Exception {
		Packet packet = new Packet();
		packet.version = Integer.parseInt(new String(stream.readNBytes(3)), 2);
		packet.typeId = Integer.parseInt(new String(stream.readNBytes(3)), 2);
		
		if(packet.typeId == 4) {
			StringBuilder sb = new StringBuilder();
			byte[] pack;
			do {
				pack = stream.readNBytes(5);
				sb.append(new String(pack, 1, 4));
			} while(pack[0] != '0');
			packet.value = Long.parseLong(sb.toString(), 2);
		} else {
			if(stream.readUnsignedByte() == '0') {
				int mark = stream.available() - Integer.parseInt(new String(stream.readNBytes(15)), 2) - 15;
				while(stream.available() > mark) {
					packet.packets.add(readPacket(stream));
				}
			} else {
				int length = Integer.parseInt(new String(stream.readNBytes(11)), 2);
				for(int i = 0; i < length; i++) {
					packet.packets.add(readPacket(stream));
				}
			}
		}
		
		return packet;
	}
	
	private static long sumPacketVersions(Packet packet) {
		long result = packet.version;
		
		for(Packet p : packet.packets) {
			result += sumPacketVersions(p);
		}
		
		return result;
	}
	
	private static long calculateExpression(Packet packet) {
		long result = 0;
		
		result = packet.typeId == 4 ? packet.value:calculateExpression(packet.packets.get(0));
		for(int i = 1; i < packet.packets.size(); i++) {
			long value = calculateExpression(packet.packets.get(i));
			switch(packet.typeId) {
				case 0 -> result += value;
				case 1 -> result *= value;
				case 2 -> result = Math.min(result, value);
				case 3 -> result = Math.max(result, value);
				case 5 -> result = result > value ? 1:0;
				case 6 -> result = result < value ? 1:0;
				case 7 -> result = result == value ? 1:0;
			}
		}
		
		return result;
	}
	
	// Solve: 37 min
	public static long partOne(Packet packet) throws Exception {
		return sumPacketVersions(packet);
	}
	
	// Solve 5 min
	public static long partTwo(Packet packet) throws Exception {
		return calculateExpression(packet);
	}
}
